pages/slice/language-reference/slice-grammar.md

```diff
@@ -2,14 +2,572 @@
+
+Submodules are modules that are logically contained within another module (called the _parent_).
+Slice definitions within a submodule can reference types in parent modules without qualification.
+
+```slice
+module A // Declares a top-level module named `A`.
+struct AStruct {}
+```
+
+```slice
+module A::B // Declares a submodule named `B` that is contained within module `A`.
+typealias T = AStruct // Don't need to qualify as `A::AStruct`.
+```
+
+All slice definitions must be contained within modules - Slice definitions cannot exist at 'global' scope.
+If a Slice file contains no Slice definitions, it is legal to omit the module declaration.
+Otherwise, if a Slice file contains any Slice definitions, a module declaration is required,
+and any Slice definitions in the remainder of the file will be contained within the declared module.
+
+Module declarations must come before any Slice definitions and must come after any [file attributes][attribute] or [mode declarations][mode-declaration].
+There can be at most one module declaration per Slice file.
+
+```ebnf {% showTitle=false %}
+SliceFile
+    : SliceFilePrelude Module? Definition*
+    ;
+```
+
+It is legal for multiple Slice files to declare the same module.
+Slice definitions defined in those files will be in the same module, even if they're in different Slice files.
+
+For additional information on modules, see the [module][module-guide] page.
+
+### Primitive types
+
+Slice supports 17 primitive types (although `AnyClass` is only valid in [`Slice1`][compilation-mode-guide] mode).
+Primitive types are always available: they're defined in any scope of any Slice file.
+They can be referenced with their keywords:
+
+```ebnf {% showTitle=false %}
+Primitive
+    : "bool"      | "int8"  | "uint8"  | "int16"    | "uint16"    | "int32"   | "uint32"  | "varint32"
+    | "varuint32" | "int64" | "uint64" | "varint62" | "varuint62" | "float32" | "float64" | "string"
+    | "AnyClass"
+    ;
+```
+
+For additional information on primitive types, see the [primitive type][primitive-type-guide] page.
+
+### Collection types
+
+Slice supports 2 built-in collection types: sequences and dictionaries.
+Both types can be referenced with their respective keywords: `Sequence` and `Dictionary`,
+
+```ebnf {% showTitle=false %}
+Sequence
+    : "sequence" "<" TypeRef ">"
+    ;
+
+Dictionary
+    : "dictionary" "<" TypeRef "," TypeRef ">"
+    ;
+```
+
+Since they are are generic over the types of elements they contain, you must also specify their type arguments:
+
+```slice
+Sequence<bool>
+Dictionary<uint8, string>
+```
+
+Since dictionary keys must be comparable, there are restrictions on which types can be used for them.
+For additional information on sequences and dictionaries, see the [sequence][sequence-guide] and [dictionary][dictionary-guide] pages.
+
+### Struct types
+
+Struct definitions consist of the `struct` keyword, followed by an [identifier], and then the struct's body.
```

Comment:
The main advantage (for me) is that we can **link to things** in the textual version, which makes it easier to navigate the reference. Say you're reading about structs and want to know about fields.
It would be awesome having the EBNF tokens (in the code block) be clickable links, but this is hard to implement.

Also the textual version **does provide some additional information**.
The grammar rule `UndelimitedList` doesn't communicate that fields may be optionally separated by commas.
`Prelude` doesn't communicate that this can be attributes or a doc comment.
etc. You _can_ get this information by navigating to those rules, but mentioning them here makes it easier to read.
A reader doesn't need to hunt around for this extra information.

This is also a **standard practice**. It's formatted differently, but you can see the same idea in the C# reference:
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes#1521-general

Reply:
I very much disagree. There is no point in just reading the EBNF grammar for you. If there is relevant additional information, sure, give this _additional_ information.

For example:
> Struct definitions consist of the struct keyword, followed by an [identifier], and then the struct's body. Struct bodies consist of a list of [fields][field] wrapped in a pair of braces. 

I don't see any information above that goes beyond a plain reading of the EBNF grammar.

As for the links, I would simply add:
See also: link1, link2, link3.

Reply:
> I very much disagree.

What are you disagreeing with here?
Having these text blocks at all?
Or one of the reasons given by @externl or I?

----

> In any of the above NOT expressed by the ebnf grammar?
> I don't see any information above that goes beyond a plain reading of the EBNF grammar.

I already answered this in my prior comment. There _is_ information that goes beyond a plain reading:
> The grammar rule UndelimitedList doesn't communicate that fields may be optionally separated by commas.
Prelude doesn't communicate that this can be attributes or a doc comment.

The grammar rule does not convey this information, but the sentences do. For reference:
```
Struct
    : Prelude "compact"? "struct" Identifier "{" UndelimitedList<Field> "}"
    ;
```

Putting this information in context, with written sentences makes it easier to read,
instead of just linking to other parts of the specification.

I would also think a textual version of rules would be more digestible to some readers then
"Here's a grammar rule in an unfamiliar syntax"